core/field_variable.js

```diff
@@ -348,6 +350,14 @@ Blockly.FieldVariable.prototype.onItemSelected = function(menu, menuItem) {
       // Delete variable.
       workspace.deleteVariableById(this.variable_.getId());
       return;
+    } else if (id == 'CREATE_VARIABLE') {
+      // Create a variable.
+      var that = this;
+      Blockly.Variables.createVariableButtonHandler(workspace, function(text) {
+        var variable = workspace.getVariable(text);
```

Comment:
What is the type of the variable?

Consider only having this option if the variable field can only hold a single type, and set the type of the new variable to the only type allowed by the field.

Reply:
Right, yes that doesn't cover the dynamic variable scenario. I'll make it only show for variables of no type for now.

Reply:
Not sure if this is a problem with the dynamic_variables stuff but I would have thought this.defaultType_ would be set to non null in a dynamic variable. Tried Blockly mainline and same issue there. Is my assumption wrong? 
I would have to resort to reading the block type to figure out if I should show this option otherwise.

Reply:
Is it null or the empty string? [This](https://github.com/google/blockly/blob/master/core/field_variable.js#L255) code seems to indicate it would be the empty string.

Checking the block type is not the right move because fields can be on any type of block.  You can check the variableTypes array instead.

Reply:
this.defaultType_ is empty string and this.variableTypes is null, even in Blockly mainline typed variable playground.